src/west/commands/project.py

```diff
@@ -726,6 +727,18 @@ def do_run(self, args, user_args):
 # Private helper routines.
 #
 
+def _clean_west_refspace(project):
+    # Cleanup fetch if sha and everything != current revision.
+    west_refs_cmd = ('for-each-ref --format="%(refname)" -- ' +
+                     WEST_REF + '**')
+    cp = project.git(west_refs_cmd, capture_stdout=True)
+    west_references = cp.stdout.decode('utf-8').strip()
+
+    for ref in west_references.splitlines():
+        if ref != WEST_REF + project.revision:
+            update_ref_cmd = 'update-ref -d ' + ref
+            project.git(update_ref_cmd)
```

Comment:
@marc-hb @mbolivar 
This will clean out any rererences from the `refs/west` ref space, unless revision is a non-SHA and identical to current revision in the manifest file.

As all references west are removed, it means that all SHAs not referenced from elsewhere (as example a tag or local branch) will be a loose object.
Loose objects are per default pruned after two weeks.
https://git-scm.com/docs/git-gc 
I believe, that this approach should be safe, as two weeks is most likely long enough time for objects not referenced elsewhere.

But if desired, it could be considered to make an entry in reflog.
The disadvantage of that, is that it starts making a lot of `west` caused entries, which might be annoying for some users (personally I prefer that reflog is populated mainly to `git` operations I have made).

So I propose to wait for user feedback before populating the reflog.

The following code presents how the reflog could be populated.
Reflog default expires after 90 days.
```suggestion
            update_ref_cmd = 'update-ref --create-reflog -m"west entry: Updating HEAD to:"  HEAD ' + ref
            project.git(update_ref_cmd)
            update_ref_cmd = 'update-ref -d ' + ref
            project.git(update_ref_cmd)
```

Reply:
> This could be annoying for some users (personally I prefer that reflog is populated mainly to git operations I have made).

There is one reflog per branch. I didn't suggest to use the reflog of some regular user branch and while I haven't looked at it in detail, the `--create-reflog` in your sample code suggests you wouldn't be polluting any regular reflog either.

Using a reflog would offer some more power and flexibility: unlike loose objects it would let users find anything recent (just like in a regular reflog), it would let users override the default expiration time, maybe others.

Reply:
> > This could be annoying for some users (personally I prefer that reflog is populated mainly to git operations I have made).
> 
> There is one reflog per branch. I didn't suggest to use the reflog of some regular user branch and while I haven't looked at it in detail, the `--create-reflog` in your sample code suggests you wouldn't be polluting any regular reflog either.
> 
> Using a reflog would offer some more power and flexibility: unlike loose objects it would let users find anything recent (just like in a regular reflog), it would let users override the default expiration time, maybe others.

Good point.
Will look a bit more into the reflog potential.
I haven't played much with reflogs in this way, mostly used reflog when I have done something I regret and want to go back to a place I don't remember ;-)

Reply:
Now creating a reflog entry in the `manifest-rev` as this reflog is actually owned by `west`.
In order to keep the manifest reflog a little easier to examine, then all entries created during the fetch part of `west update`, will have the message: `west fetch: entry for {ref}`, where `{ref}` is of the form: `refs/west/<name>`.

Reply:
Interesting, thanks! I will try to take another look at this on Monday morning my time.

Reply:
> Now creating a reflog entry in the manifest-rev as this reflog is actually owned by west.

You mean this is already done, right? I see a reflog for manifest-rev in my existing west-managed git repos. For example, `git -C ncs/zephyr reflog manifest-rev` works for me.

> In order to keep the manifest reflog a little easier to examine, then all entries created during the fetch part of `west update`, will have the message: `west fetch: entry for {ref}`, where `{ref}` is of the form: `refs/west/<name>`.

It looks like that's true only if the user specifically does something like enable core.logAllRefUpdates, right?

```
LOGGING UPDATES
       If config parameter "core.logAllRefUpdates" is true and the ref is one
       under "refs/heads/", "refs/remotes/", "refs/notes/", or the symbolic
       ref HEAD; or the file "$GIT_DIR/logs/<ref>" exists then git update-ref
       will append a line to the log file "$GIT_DIR/logs/<ref>" (dereferencing
       all symbolic refs before creating the log name) describing the change
       in ref value. 
```

Or did you mean to push a change with `--create-reflog` added?

Reply:
> > Now creating a reflog entry in the manifest-rev as this reflog is actually owned by west.
> 
> You mean this is already done, right? I see a reflog for manifest-rev in my existing west-managed git repos. For example, `git -C ncs/zephyr reflog manifest-rev` works for me.
>

Correct, a reflog for manifest-rev is already there with entries on each `git update-ref`.
What I meant, was that a useable entry is now created.
Currently `west update` will create entries such as:
```
a6a98b7854 (HEAD, refs/west/pull/240/head, manifest-rev) manifest-rev@{0}: 
b687d5d022 manifest-rev@{1}: 
a6a98b7854 manifest-rev@{2}:
b687d5d022 manifest-rev@{3}: 
69a4c73530 manifest-rev@{4}: 
b687d5d022 manifest-rev@{5}: 
....
``` 
when running `west update` and `git update-ref` is executed in background.

> > In order to keep the manifest reflog a little easier to examine, then all entries created during the fetch part of `west update`, will have the message: `west fetch: entry for {ref}`, where `{ref}` is of the form: `refs/west/<name>`.
> 
> It looks like that's true only if the user specifically does something like enable core.logAllRefUpdates, right?
> 

No, by actually specifying the `refs/west/<name>` we allow a user that is examine the reflog to actually see that `<shaA>` in the reflog originated from `pull/<no>/head` and `<shaB>` is from `<branch>`, and `shaC` is from same `pull/<no>/head` as `<shaA>`, which could indicate a forced push. 

So now there could be:
```
a6a98b7854 manifest-rev@{2}: west fetch: entry for refs/west/pull/240/head
eedc25e5b2 manifest-rev@{6}: west fetch: entry for refs/west/v2.0.99-ncs1-branch
7e64ac737d manifest-rev@{7}: west fetch: entry for refs/west/v1.14.99-ncs2-branch
1e34490406 manifest-rev@{8}: west fetch: entry for refs/west/nrf91
c373e5578c manifest-rev@{9}: west fetch: entry for refs/west/master
902e57651b manifest-rev@{15}: west fetch: entry for refs/west/pull/240/head <--- A forced push is made later a manifest-rev@{2}
```